### PR TITLE
The grand reopening Part 3: Add basic DEM

### PR DESCRIPTION
**Pull request by svchb**
278239988-cd069d69-1d41-4a98-9484-1e9fc033403c

### CODE REVIEW

**Comment by svchb** [comment]
Add to news.md and add Docs.

**Comment by svchb** [comment]
The test still fail because this has a breaking change to the dummy particles system constructor. I want to discuss this first before making the adjustments.

**Review by efaulhaber — CHANGES_REQUESTED** [review]
Very cool!
I didn't find the corresponding equations in the two references, so I didn't look at the equations itself. Can you please add a reference for this exact interaction? Or specify where exactly one can find it in the references.

**Comment by efaulhaber** [comment]
But where did you get the equations from?

**Comment by svchb** [comment]
I did not use a paper.
Anyway here is a overview of some of the equations:
https://www.sciencedirect.com/science/article/pii/S0009250903005414

**Comment by svchb** [comment]
I assume frictionless contact though

File: src/schemes/solid/discrete_element_method/rhs.jl [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                                      particle)
+end
+
+@inline function position_correction!(neighbor_system::BoundaryDEMSystem, u_particle_system,
+                                      overlap, normal, particle)
+    for i in 1:ndims(neighbor_system)
+        # Position correction to prevent penetration
+        u_particle_system[i, particle] -= 0.5 * overlap * normal[i]
+    end
+end
```
Comment (efaulhaber) [inline src/schemes/solid/discrete_element_method/rhs.jl:106]: You can't modify `u` inside a rhs call. This is not allowed by OrdinaryDiffEq.

File: src/schemes/solid/discrete_element_method/system.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+   captured by standard DEM interactions, such as electromagnetic forces or user-defined perturbations.
+- `damping_coefficient=0.0001`: Set a damping coefficient for the collision interactions.
+"""
+struct DEMSystem{NDIMS, ELTYPE <: Real, ST} <: SolidSystem{NDIMS}
+    initial_condition   :: InitialCondition{ELTYPE}
+    mass                :: Array{ELTYPE, 1}     # [particle]
+    radius              :: Array{ELTYPE, 1}     # [particle]
+    elastic_modulus     :: ELTYPE
+    poissons_ratio      :: ELTYPE
+    kn                  :: ELTYPE               # Normal stiffness
```
Comment (efaulhaber) [inline src/schemes/solid/discrete_element_method/system.jl:29]: How about calling the variable `normal_stiffness`? That's a lot more descriptive.

File: src/schemes/solid/discrete_element_method/system.jl [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+- `kn`: Normal stiffness coefficient for particle-particle and particle-wall contacts.
+- `elastic_modulus`: Elastic modulus for this particle system.
+- `poissons_ratio`: Poisson ratio for this particle system.
+
+# Keyword
+- `acceleration`: Global acceleration vector applied to the system, such as gravity. Specified as
+   an `SVector` of length `NDIMS`, with a default of zero in each dimension.
+- `source_terms`: Optional; additional forces or modifications to particle dynamics not
+   captured by standard DEM interactions, such as electromagnetic forces or user-defined perturbations.
+- `damping_coefficient=0.0001`: Set a damping coefficient for the collision interactions.
```
Comment (efaulhaber) [inline src/schemes/solid/discrete_element_method/system.jl:21]: 1. Indentation
2. `Keywords`
3. Trying to get rid of the misleading use of the word "system" for both physical systems and code structures. Also, "tracks the motion" sounds weird IMO.
```suggestion
    DEMSystem(initial_condition, kn, elastic_modulus, poissons_ratio;
              damping_coefficient=0.0001, acceleration=ntuple(_ -> 0.0, NDIMS), source_terms=nothing)

System for particles of a granular material.
The Discrete Element Method (DEM) is used to simulate the dynamics of the material.
Creates a Discrete Element Method (DEM) system for simulating the dynamics of particulate systems.
Each particle is considered discrete, as opposed to continuous methods like SPH,
and interactions between particles are modeled using forces such as normal contact stiffness.

# Arguments
- `initial_condition`: Initial condition of the system, encapsulating the initial positions,
                       velocities, masses, and radii of particles.
- `kn`:                Normal stiffness coefficient for particle-particle and particle-wall contacts.
- `elastic_modulus`:   Elastic modulus for this particle system.
- `poissons_ratio`:    Poisson ratio for this particle system.

# Keywords
- `acceleration`: Global acceleration vector applied to the system, such as gravity. Specified as
                  an `SVector` of length `NDIMS`, with a default of zero in each dimension.
- `source_terms`: Optional; additional forces or modifications to particle dynamics not
                  captured by standard DEM interactions, such as electromagnetic forces or user-defined perturbations.
- `damping_coefficient=0.0001`: Set a damping coefficient for the collision interactions.
```

**Review by efaulhaber — CHANGES_REQUESTED** [review]

File: examples/dem/rectangular_tank_2d.jl [hunk 4]
```diff
@@ -21,20 +17,15 @@ tank_height = 4.0
...
 
 # ==========================================================================================
 # ==== Systems
 
-# let them fall
+# Move the rocks up to let them fall
 tank.fluid.coordinates[2, :] .+= 0.5
-solid_system = DEMSystem(tank.fluid, 2 * 10^5, acceleration=(0.0, gravity))
-boundary_system = BoundaryDEMSystem(tank.boundary, boundary_model)
+solid_system = DEMSystem(tank.fluid, 2 * 10e5, 10e9, 0.3, acceleration=(0.0, gravity))
```
[the remark this answers is not in the document]
Comment (svchb) [reply examples/dem/rectangular_tank_2d.jl:27]: Well its a system of solid spherical rocks

File: src/schemes/solid/discrete_element_method/rhs.jl (hunk 1 above)
Comment (LasNikas) [reply src/schemes/solid/discrete_element_method/rhs.jl:106]: Yes, you need the callback from #440

File: src/general/initial_condition.jl [hunk 5]
```diff
@@ -163,8 +164,10 @@ struct InitialCondition{ELTYPE}
             masses = mass_fun.(coordinates_svector)
         end
 
+        radius = 0.5 * particle_spacing * ones(length(masses))
```
[the remark this answers is not in the document]
Comment (svchb) [reply src/general/initial_condition.jl:167]: This is of course generally not true you could have different sized particles but then I would need to make more changes while there is no initialization allowing to set this up correctly at the moment.

File: docs/src/systems/dem.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# [Discrete Element Method](@id dem)
+The Discrete Element Method (DEM) is a computational technique widely used in physics, engineering,
+and applied mathematics for simulating the dynamics of systems of particles. This method allows
```
[the remark this answers is not in the document]
Comment (efaulhaber) [reply docs/src/systems/dem.md]: How about
```
The Discrete Element Method (DEM) is a computational technique widely used in physics, engineering,
and applied mathematics for simulating the mechanical behavior of granular materials, such as powders, sand, soil, or rock,

File: src/general/initial_condition.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (efaulhaber) [reply src/general/initial_condition.jl:167]: Then why not leave the initial condition the way it is and just initialize the radius of the DEM system with half the particle spacing of the initial condition?

**Review by efaulhaber — CHANGES_REQUESTED** [review]

File: src/general/initial_condition.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (svchb) [reply src/general/initial_condition.jl:167]: This has to be added here anyway in the future? Or do we want multiple initial condition types?

[the remark this answers is not in the document]
Comment (efaulhaber) [reply src/general/initial_condition.jl:167]: I suggest we leave the initial condition as is, and think about modifying it once that feature will actually be used.

[the remark this answers is not in the document]
Comment (svchb) [reply src/general/initial_condition.jl]: ok I have removed it

**Review by efaulhaber — APPROVED** [review]

**Review by efaulhaber — APPROVED** [review]

**Pull request merged**